File: mobility/Motors.py
#!/usr/bin/env python

#Richard G. Hirst's ServoBlaster available at https://github.com/richardghirst/PiBits/tree/master/ServoBlaster
#pulsewidth: 1000microseconds for full speed ahead, 
#            2000microseconds for full speed reverse, 
#            1500microseconds for idling and 
#            200microseconds for brake
#note that Servoblast provides 10microsecond pulse widths, so writing "100" to /dev/servoblaster gives full speed ahead


import time
import os
import subprocess
import re
import math

class Motors:
    "Motor Drivers using Servoblaster for Vex Robotics 393 motors via Vex Motor Controller 29's"
    def __init__(self):
        self.leftMotor = 2 #pin 2 in servoblaster, GPIO 18 and Pin P1-12 on the RPi
        self.rightMotor = 1 #pin 1 in servoblaster, GPIO 17 and Pin P1-11 on the RPi
        speed = 0 #from 0-50 in arbitrary units <>TODO: TRANSLATE TO ACTUAL SPEED


    #make sure Servoblaster daemon is running
    def setup(self):
        s = subprocess.Popen(["ps","axw"],stdout=subprocess.PIPE)
        for x in s.stdout:
            if re.search("servod",x):
                print "Servoblaster Daemon is already running"
            else:
                cmd = "sudo ./ServoBlaster/servod"
                os.system(cmd)
                break

    #close Servoblaster daemon
    def teardown(self):
        s = subprocess.Popen(["ps","axw"],stdout=subprocess.PIPE)
        for x in s.stdout:
            if re.search("servod",x):
                print "Servoblaster Daemon has been shut down"
                os.system("sudo killall servod")
                break

    #output to motor control file
    def setMotors(self,leftMotorSpeed,rightMotorSpeed):
        f = open("/dev/servoblaster", "w")
        f.write(str(self.leftMotor) + "=" + str(150 + leftMotorSpeed) + "\n")
        f.write(str(self.rightMotor) + "=" + str(150 - rightMotorSpeed) + "\n")
        f.close()

    #move a set distance (positive for forward, negative for backward)
    def goDistance(self, speed, distance):
        if (distance > 0):
            self.setMotors(speed,speed)
        else:
            self.setMotors(-speed,-speed)
        time.sleep(math.fabs(distance))
        self.setMotors(0,0)

    #rotate a set angle (positive for counterclockwise, right for clockwise)
    def rotateAngle(self, speed, angle):
        if (angle > 0):
            self.setMotors(-speed,speed)
        else:
            self.setMotors(speed,-speed)
        time.sleep(math.fabs(angle))
        self.setMotors(0,0)

    #Go straight ahead
    def goForward(self, speed):
        self.setMotors(speed,speed)

    #Go straight backwards
    def goBackward(self, speed):
        self.setMotors(-speed,-speed)

    #Turn left while moving forward
    def turnLeft(self, speed):
        self.setMotors(speed/3,speed)

    #Turn right while moving forward
    def turnRight(self, speed):
        self.setMotors(speed,speed/3)

    #Pivot about left wheel
    def pivotLeft(self, speed):
        self.setMotors(0,speed)

    #Pivot about right wheel
    def pivotRight(self, speed):
        self.setMotors(speed,0)

    #Rotate left (CCW) about center of rotation
    def rotateLeft(self, speed):
        self.setMotors(-speed,speed)

    #Rotate right (CW) about center of rotation
    def rotateRight(self, speed):
        self.setMotors(speed,-speed)

    #Rotate right (CW) about center of rotation
    def brake(self):
        self.setMotors(0,0)

motors = Motors()
motors.setup()
timestep = 0.1 # seconds

# test motor commands
for i in range(50):
    motors.goForward(i)
    time.sleep(timestep)

for i in range(50):
    motors.goBackward(i)
    time.sleep(timestep)

for i in range(50):
    motors.turnRight(i)
    time.sleep(timestep)

for i in range(50):
    motors.turnLeft(i)
    time.sleep(timestep)

for i in range(50):
    motors.pivotRight(i)
    time.sleep(timestep)

for i in range(50):
    motors.pivotLeft(i)
    time.sleep(timestep)

for i in range(50):
    motors.rotateRight(i)
    time.sleep(timestep)

for i in range(50):
    motors.rotateLeft(i)
    time.sleep(timestep)

#move back and forth
motors.goForward(30)
time.sleep(1)
motors.goBackward(30)
time.sleep(1)
motors.brake()

#move distance and angle
motors.goDistance(15,1)
motors.goDistance(15,-1)
motors.rotateAngle(15,3)
motors.rotateAngle(15,-3)

motors.teardown()

File: mobility/Motors.cpp
#include <time.h>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <stdlib.h>
#include <cmath>
#include "Motors.h"

#define MICROSEC2SEC 1000000

using namespace std;

Motors::Motors(){
    leftMotor = '2'; //pin 2 in servoblaster, GPIO 18 and Pin P1-12 on the RPi
    rightMotor = '1'; //pin 1 in servoblaster, GPIO 17 and Pin P1-11 on the RPi
    if (0 == system("pidof 0x servod > /dev/null")){
        cout << "ServoBlaster Daemon is already running!" << endl;
    } else{
        system("sudo ./ServoBlaster/servod");
    }
}

Motors::~Motors(){
    if (0 == system("pidof 0x servod > /dev/null")){
        system("sudo killall servod");
        cout << "ServoBlaster Daemon has been shut down!" << endl;
    }   
}

void Motors::setMotors(int leftMotorSpeed, int rightMotorSpeed){
    ofstream f;
    f.open("/dev/servoblaster", ios::trunc | ios::out);
    f << this->leftMotor << "=" << 150 + leftMotorSpeed << '\n';
    f << this->rightMotor << "=" << 150 - rightMotorSpeed << '\n';
    f.close();
}

//Go straight ahead
void Motors::goForward(int speed){
    this->setMotors(speed,speed);
}

//Go straight backwards
void Motors::goBackward(int speed){
    this->setMotors(-speed,-speed);
}

//Turn left while moving forward
void Motors::turnLeft(int speed){
    this->setMotors(speed/3,speed);
}

//Turn right while moving forward
void Motors::turnRight(int speed){
    this->setMotors(speed,speed/3);
}
//Pivot about left wheel

void Motors::pivotLeft(int speed){
    this->setMotors(0,speed);
}

//Pivot about right wheel
void Motors::pivotRight(int speed){
    this->setMotors(speed,0);
}

//Rotate left (CCW) about center of rotation
void Motors::rotateLeft(int speed){
    this->setMotors(-speed,speed);
}

//Rotate right (CW) about center of rotation
void Motors::rotateRight(int speed){
    this->setMotors(speed,-speed);
}

//Rotate right (CW) about center of rotation
void Motors::brake(){
    this->setMotors(0,0);
}

/*int main(){
    Motors motors;
    double timestep = 0.1; // seconds

    // test motor commands
    for (int i = 0; i<50; i++){
        motors.goForward(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.goBackward(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.turnRight(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.turnLeft(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.pivotRight(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.pivotLeft(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.rotateRight(i);
        usleep(MICROSEC2SEC *timestep);
    }

    for (int i = 0; i<50; i++){
        motors.rotateLeft(i);
        usleep(MICROSEC2SEC *timestep);
    }

    //move back and forth
    motors.goForward(30);
    usleep(MICROSEC2SEC *1);
    motors.goBackward(30);
    usleep(MICROSEC2SEC *1);
    motors.brake();

    //move distance and angle
    motors.goDistance(15,1);
    motors.goDistance(15,-1);
    motors.rotateAngle(15,3);
    motors.rotateAngle(15,-3);
    motors.goDistance(20,1);
    motors.goDistance(20,-1);
    motors.rotateAngle(20,1);
    motors.rotateAngle(20,-2);
    motors.rotateAngle(20,1);

    return 0;
}*/
